GitExtUtils/HexString.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System.Text;
+
+namespace GitCommands
+{
+    public static class HexString
+    {
+        /// <summary>
+        /// Converts a byte array into its hex representation.
+        /// </summary>
+        public static string FromByteArray(byte[] data, bool lowerCase)
+        {
+            if (data is null)
+            {
+                return null;
+            }
+
+            if (data.Length == 0)
+            {
+                return string.Empty;
+            }
```

Comment:
Why return different values? Why is it significant?

Reply:
Because that's what I would expect using a method like this.
I wouldn't want it to crash (that's why the null check is there).

And if I provide an input (even though it's an empty array) I wouldn't expect `null` as an answer.
If I would drop the second condition the result would still be the same (an empty string) but generated through the StringBuilder.

The second case is basically just a "shortcut".

Reply:
In `GravatarProvider` both `null` and `""` become `""`.
`TemplateFormatter` that indirectly uses `FromByteArray` will eventually turn `null` into `""` as well. Along the way it is totally not obvious how/if `null`s are handled, it is totally opaque to a caller.

I can't really see a reason to return `null` (and risk getting a NRE). Convince me otherwise :)

Reply:
_"Convince me otherwise :)"_

Doesn't sound like someone wants to be convinced. My guess is that is just would result in an argument with an double down effect ^^

`HexString` has nothing to do with `GravatarProvider` (in the same way the priviously existing explicit upper-case/lower-case parameter) had nothing to do with the `GravatarProvider`. So if we design a class in a way that it's custom tailored to work with a specific other class (in this case tailoring `HexString` for the use inside `GravatarProvider`) doesn't sound like good design decisions to me.

Regarding NRE - there is no more or less risk to NRE because the byte array was clearly null in the first place.

A mathematical analogy, a method that calculates the product:

```csharp
int Multiply(params int[] list)
{
    if (list is null)
    {
        // we return null because we don't want this method to throw, we don't have to.
        // if the consumer passes null the only reasonable answer he can expect is either an ArgumentNullException
        // or null as a return value. Returning something else would IMHO be a big surprise
        return null;
    }

    if (list.Length == 0)
    {
        // Empty product: (https://en.wikipedia.org/wiki/Empty_product)
        // Note that this is just a shortcut an removing this case would also cause the method to return 1
        // Also note that a missing argument (null) is not the same same as an empty input set.
        return 1;
    }
    
    var product = 1;

    for (int i = 0; i < list.Length; i++)
    {
        product *= list[i];
    }

    return product ;
}
```

Non the less, I'm not the maintainer here :-) I can change it if you want me to.

Reply:
> return null;

What does the method return in this case?
Execution result or execution status.
I think this is the key question.

Reply:
The result. I'm not sure what you mean by execution status because the method is not asynchronous.

Reply:
As I understood from your example, you want to return a status so as not to throw an exception. We mix the two concepts. If we need statust here, it is better to use out parameter.

In the current case, I suggest that this method always works and returning an empty string.

Reply:
✔ done